services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceParentHelixAdmin.java

```diff
@@ -2513,12 +2494,48 @@ public void updateStore(String clusterName, String storeName, UpdateStoreQueryPa
             ErrorType.BAD_REQUEST);
       }
 
-      final boolean writeComputeJustEnabled =
+      // When this config is true, it means partial update is enabled by request.
+      boolean partialUpdateJustEnabled =
           writeComputationEnabled.orElse(false) && !currStore.isWriteComputationEnabled();
-      if (writeComputeJustEnabled) {
+      if (partialUpdateJustEnabled) {
         // Dry-run generating Write Compute schemas before sending admin messages to enable Write Compute because Write
         // Compute schema generation may fail due to some reasons. If that happens, abort the store update process.
         addWriteComputeSchemaForStore(clusterName, storeName, true);
+      } else {
+        /**
+         * If a store: (1) Is being converted to hybrid (2) Is not partial update enabled for now. (3) Does not change
+         * partial update config in this request.
+         * It means partial update is not enabled, and there is no explict intention to change it. In this case, we will
+         * look up cluster config and based on the replication policy to enable partial update implicitly.
+         */
+        final boolean shouldEnablePartialUpdateBasedOnClusterConfig =
```

Comment:
Currently, the logic is a little weird and I think it is also buggy..
For example, if someone disables partial update, and it will execute the `else` branch, which may re-enable partial update..

We should have a section to infer whether we should enable partial update or not.
```
If partial update param is explicitly set, Controller should follow it.
If not, Controller will use the cluster config to decide whether to enable partial update or not when `storeBeingConvertedToHybrid` is `true`.

When partial update needs to be enabled for the first time because of the above logic, Controller will try to generate update schemas.
```
This might be more clear.

Reply:
Yeah, I do agree about this, and I have a vague idea to refactor the parent controller update store logic to scatter all the complicated param into pieces and execute them in topological order. Let me do it for partial update first to clean this up incrementally...

Reply:
I refactored a bit and move it into util class and add java docs to document the logics. Please feel free to take a 2nd look.